File: tools/minimal/backend/read_shared_context.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
read_shared_context - 共享上下文池读取工具
读取 analyze_project 等工具写入的共享数据（app_data 表，全局共享，
不按对话隔离），供流程图每个节点/任意对话共同使用。
"""
import json
from tools.coding.backend.base import ToolContext

TOOL_NAME = 'read_shared_context'
CATEGORY = 'analyze_project'  # 与 analyze_project 共用同一类别，读同一份数据

VALID_KEYS = ('project_analysis',)


def _get_conn(ctx):
    """在 db_lock 保护下获取数据库连接（调用方负责 close）"""
    with ctx.db_lock:
        conn = ctx.get_db()
    return conn


def handle(body, ctx):
    try:
        key = body.get('key') or 'project_analysis'
        if key not in VALID_KEYS:
            return ctx.send_json({
                'ok': False,
                'error': f'不支持的 key: {key}，可用: {list(VALID_KEYS)}',
            })
        part = body.get('part') or 'summary'   # summary/files/routes/mermaid/all
        limit = min(int(body.get('limit') or 200), 1000)

        conn = _get_conn(ctx)
        try:
            row = conn.execute(
                'SELECT value FROM app_data WHERE category=? AND key=?',
                [CATEGORY, key]).fetchone()
        finally:
            conn.close()
        if not row:
            return ctx.send_json({
                'ok': False,
                'error': '尚无分析数据，请先调用 analyze_project（action=analyze）。',
            })
        data = json.loads(row['value'])

        if part == 'summary':
            out = {'root': data['root'], 'analyzed_at': data['analyzed_at'],
                   'summary': data['summary']}
        elif part == 'mermaid':
            out = {'root': data['root'], 'mermaid': data['mermaid']}
        elif part == 'files':
            out = {'root': data['root'],
                   'total': len(data['files']),
                   'files': data['files'][:limit]}
        elif part == 'routes':
            out = {'root': data['root'],
                   'total': len(data['routes']),
                   'routes': data['routes'][:limit]}
        else:  # all
            out = data

        return ctx.send_json({'ok': True, 'part': part, 'data': out})
    except Exception as e:
        return ctx.send_json({'ok': False, 'error': str(e)})

File: tools/minimal/frontend/read_shared_context.js
// ========== read_shared_context.js - 工具定义 ==========
// 朱峰社区智能体无限 5.0.8 - 共享上下文池读取

window.Tools = window.Tools || { allTools: {} };
if (!window.Tools.allTools['read_shared_context']) {
    window.Tools.allTools['read_shared_context'] = {
        type: 'function',
        function: {
            name: 'read_shared_context',
            description: '读取共享上下文池。analyze_project 等工具写入的全局共享数据（不按对话隔离，流程图所有节点、所有对话都可读）。key 目前支持 project_analysis（项目分析结果）。part 可选 summary=摘要（默认）、mermaid=流程图、files=文件列表、routes=路由列表、all=全部。limit 限制列表返回条数（默认 200，最大 1000）。',
            parameters: {
                type: 'object',
                properties: {
                    key: {
                        type: 'string',
                        enum: ['project_analysis'],
                        description: '要读取的数据键（默认 project_analysis）'
                    },
                    part: {
                        type: 'string',
                        enum: ['summary', 'mermaid', 'files', 'routes', 'all'],
                        description: '读取哪部分（默认 summary）'
                    },
                    limit: {
                        type: 'integer',
                        description: '列表类数据的最大返回条数（默认 200，最大 1000）'
                    }
                },
                required: []
            }
        }
    };
}
